src/mcts/search.h

```diff
@@ -282,50 +301,143 @@ class SearchWorker {
+    std::vector<std::unique_ptr<int[]>> vtp_buffer;
+  };
 
-  NodeToProcess PickNodeToExtend(int collision_limit);
-  void ExtendNode(Node* node, int depth);
-  bool AddNodeToComputation(Node* node, bool add_if_cached, int* transform_out);
+  void PickNodesToExtend(int collision_limit);
+  void PickNodesToExtendTask(Node* starting_point, int collision_limit,
+                             int base_depth,
+                             const std::vector<Move>& moves_to_base,
+                             std::vector<NodeToProcess>* receiver,
+                             TaskWorkspace* workspace);
+  void ProcessPickedTask(int batch_start, int batch_end,
+                         TaskWorkspace* workspace);
+  void ExtendNode(Node* node, int depth, const std::vector<Move>& moves_to_add,
+                  PositionHistory* history);
+  bool AddNodeToComputation(Node* node, bool add_if_cached, int* transform_out,
+                            PositionHistory* history);
   int PrefetchIntoCache(Node* node, int budget, bool is_odd_depth);
+  template <typename Computation>
   void FetchSingleNodeResult(NodeToProcess* node_to_process,
+                             const Computation& computation,
                              int idx_in_computation);
   void DoBackupUpdateSingleNode(const NodeToProcess& node_to_process);
   // Returns whether a node's bounds were set based on its children.
-  bool MaybeSetBounds(Node* p, float m, int* n_to_fix, float* v_delta, float* d_delta, float* m_delta) const;
+  bool MaybeSetBounds(Node* p, float m, int* n_to_fix, float* v_delta,
+                      float* d_delta, float* m_delta) const;
+
+  void RunTasks(int tid);
 
   Search* const search_;
   // List of nodes to process.
   std::vector<NodeToProcess> minibatch_;
   std::unique_ptr<CachingComputation> computation_;
-  // History is reset and extended by PickNodeToExtend().
+  // History is reset and extended by PrefetchIntoCache().
   PositionHistory history_;
   int number_out_of_order_ = 0;
   const SearchParams& params_;
   std::unique_ptr<Node> precached_node_;
   const bool moves_left_support_;
   IterationStats iteration_stats_;
   StoppersHints latest_time_manager_hints_;
+
+  struct PickTask {
+    int task_type;
+
+    // For task type 0 - gathering.
+    Node* start;
+    int base_depth;
+    int collision_limit;
+    std::vector<NodeToProcess> results;
+    std::vector<Move> moves_to_base;
+
+    // Task type 1 - post gather processing.
+    int start_idx;
+    int end_idx;
+
+    bool complete = false;
+    PickTask(Node* node, uint16_t depth, const std::vector<Move>& base_moves,
+             int collision_limit)
+        : task_type(0),
+          start(node),
+          collision_limit(collision_limit),
+          base_depth(depth),
+          moves_to_base(base_moves) {}
+    PickTask(int start_idx, int end_idx)
+        : task_type(1), start_idx(start_idx), end_idx(end_idx) {}
+  };
+  std::mutex picking_tasks_mutex_;
```

Comment:
With a mutex from utils/mutex.h we can have thread safety annotations. :)
But I think at this point it may make sense to switch to absl, it's pretty easy:

https://github.com/mooskagh/lc0/blob/lc2v3/subprojects/abseil.wrap
https://github.com/mooskagh/lc0/blob/lc2v3/meson.build#L545

Reply:
I chose std::mutex because I was using condition variable and didn't know about get_raw - will fix

Reply:
Fixed in 1490.